tests/php/src/Validation/ValidationURLProviderTest.php

```diff
@@ -0,0 +1,386 @@
+<?php
+
+namespace AmpProject\AmpWP\Tests\Validation;
+
+use AmpProject\AmpWP\Option;
+use AMP_Options_Manager;
+use AMP_Post_Meta_Box;
+use AMP_Theme_Support;
+use AmpProject\AmpWP\Tests\Helpers\AssertContainsCompatibility;
+use AmpProject\AmpWP\Tests\Helpers\PrivateAccess;
+use AmpProject\AmpWP\Tests\Helpers\ValidationRequestMocking;
+use AmpProject\AmpWP\Validation\ValidationURLProvider;
+use WP_Query;
+use WP_UnitTestCase;
+
+/** @coversDefaultClass ValidationURLProvider */
+final class ValidationURLProviderTest extends WP_UnitTestCase {
+	use PrivateAccess, AssertContainsCompatibility;
+
+	/**
+	 * Setup.
+	 *
+	 * @inheritdoc
+	 */
+	public function setUp() {
+		parent::setUp();
+		$this->validation_url_provider = new ValidationURLProvider();
+		add_filter( 'pre_http_request', [ ValidationRequestMocking::class, 'get_validate_response' ] );
+	}
+
+	/**
+	 * Test retreival of urls.
+	 *
+	 * @covers ::get_urls()
+	 */
+	public function test_count_urls_to_validate() {
+		$this->validation_url_provider = new ValidationURLProvider();
+
+		$number_original_urls = 4;
+
+		$this->assertEquals( $number_original_urls, count( $this->validation_url_provider->get_urls() ) );
+
+		$this->validation_url_provider = new ValidationURLProvider( 100 );
+
+		$category         = self::factory()->term->create( [ 'taxonomy' => 'category' ] );
+		$number_new_posts = 50;
+		$post_ids         = [];
+		for ( $i = 0; $i < $number_new_posts; $i++ ) {
+			$post_ids[] = self::factory()->post->create(
+				[
+					'tax_input' => [ 'category' => $category ],
+				]
+			);
+		}
+
+		/*
+		 * Add the number of new posts, original URLs, and 1 for the $category that all of them have.
+		 * And ensure that the tested method finds a URL for all of them.
+		 */
+		$expected_url_count = $number_new_posts + $number_original_urls + 1;
+		$this->assertEquals( $expected_url_count, count( $this->validation_url_provider->get_urls() ) );
+
+		$this->validation_url_provider = new ValidationURLProvider( 100 );
+
+		$number_of_new_terms        = 20;
+		$expected_url_count        += $number_of_new_terms;
+		$taxonomy                   = 'category';
+		$terms_for_current_taxonomy = [];
+		for ( $i = 0; $i < $number_of_new_terms; $i++ ) {
+			$terms_for_current_taxonomy[] = self::factory()->term->create(
+				[
+					'taxonomy' => $taxonomy,
+				]
+			);
+		}
+
+		// Terms need to be associated with a post in order to be returned in get_terms().
+		wp_set_post_terms(
+			$post_ids[0],
+			$terms_for_current_taxonomy,
+			$taxonomy
+		);
```

Comment:
Maybe assert this call worked as intended?
```suggestion
		$result = wp_set_post_terms(
			$post_ids[0],
			$terms_for_current_taxonomy,
			$taxonomy
		);
		$this->assertFalse( is_wp_error( $result ) );
```